### PR TITLE
Added support for maximum replicas per node

### PR DESCRIPTION
**Pull request by olljanat**
**- What I did**
Added support for maximum replicas per node

Second step to be able solve moby/moby#26259

**- How I did it**
Addes support for Placement parameter MaxReplicas

TODO:

- [X] Get docker/swarmkit#2758 merged.
- [X] Wait that  #38003 is merged.

### CODE REVIEW

**Comment by kolyshkin** [comment]
build fails:

> 00:07:31.935 Building: bundles/binary-daemon/dockerd-dev
> 00:08:09.664 # github.com/docker/docker/daemon/cluster/convert
> 00:08:09.664 daemon/cluster/convert/service.go:249:4: unknown field 'Maxreplicas' in struct literal of type api.Placement

**Comment by olljanat** [comment]
Sure it will until swarmkit PR have been merged that why this is marked as WIP now.

Test automation set on Moby is quite complex and I'm not able run all these tests on my dev env so that why I prefer to have PR open here which can run those tests for me.

Next step are: 
* Do requested changes swarmkit PR
* Make this PR match with that one.
* Temporarily add commit which takes swarmkit from my repo

After these build and all these tests (should) pass.

**Comment by olljanat** [comment]
Swarmkit PR was merged so I included swarkit bump to this.

Builds should pass as they was earlier when I tested to include swarmkit changes from my repo.

@cpuguy83 can you review ?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
just some minor comments, but I want to give this a spin (haven't had time for that yet).

One thing I'd like to test is the behavior if the `--max-replicas-per-node` value is updated for an existing service (e.g. change from max 3 replicas to max 2; are the existing tasks properly killed and/or re-scheduled in that case?)

**Comment by olljanat** [comment]
> One thing I'd like to test is the behavior if the --max-replicas-per-node value is updated for an existing service (e.g. change from max 3 replicas to max 2; are the existing tasks properly killed and/or re-scheduled in that case?)

Currently it looks working on way that if user only updates --max-replicas-per-node (or actually it is ```--replicas-max-per-node``` so it shorts nicely on cli help) then only service spec will be updated and extra tasks will be killed on next service update. If user also updates ```--image``` value or uses ```--force``` switch then extra one will be killed immediately (as all tasks are re-scheduled).

If we need change that logic then it would mean one more change to swarmkit.

Anyway, I will wait that #38003 is merged and implement these API versions checks on meanwhile.

**Comment by olljanat** [comment]
Temporarily included commit from #38003 will cleanup after it is merged.

File: daemon/cluster/convert/service.go [hunk 1]
```diff
@@ -246,6 +246,7 @@ func ServiceSpecToGRPC(s types.ServiceSpec) (swarmapi.ServiceSpec, error) {
 		spec.Task.Placement = &swarmapi.Placement{
 			Constraints: s.TaskTemplate.Placement.Constraints,
 			Preferences: preferences,
+			MaxReplicas: s.TaskTemplate.Placement.MaxReplicas,
```
[the remark this answers is not in the document]
Comment (olljanat) [reply daemon/cluster/convert/service.go]: These are in-place now

File: api/swagger.yaml [hunk 2]
```diff
@@ -2844,6 +2844,10 @@ definitions:
                   SpreadDescriptor: "node.labels.datacenter"
               - Spread:
                   SpreadDescriptor: "node.labels.rack"
+          MaxReplicas:
```
[the remark this answers is not in the document]
Comment (olljanat) [reply api/swagger.yaml]: Included.

**Comment by olljanat** [comment]
@thaJeztah rebased on master and included changes you requested. Can you review again?

**Comment by olljanat** [comment]
@vieux / @thaJeztah Now also tests passed. Please review.

**Comment by olljanat** [comment]
Ping @vieux @thaJeztah @cpuguy83

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by olljanat** [comment]
Rebased and fixed merge conficts.

ping @thaJeztah

**Comment by olljanat** [comment]
@yongtang I can see that you have been active here lately. Can I ask you to looks this one too?

**Review by yongtang — APPROVED** [review]
LGTM

**Comment by olljanat** [comment]
@yongtang powerpc build passed now so can you also merge? Docs updates are part of client side PR docker/cli#1410

**Comment by olljanat** [comment]
ping @thaJeztah

**Comment by olljanat** [comment]
@thaJeztah just reminder that I would like to get this one merged so I can finalize client side.

**Comment by cpuguy83** [comment]
Actually thinking we should have an integration test or two here.

**Comment by olljanat** [comment]
@cpuguy83 logic is tested on swarmkit side but I can of course add some integration tests too. Can you point me some up to date existing test which I can use as reference? (I noticed that there is multiple ways to do testing and I'm not 100% sure which way would be correct one here)

**Comment by cpuguy83** [comment]
probably integration/service/create_test.go

**Comment by olljanat** [comment]
@thaJeztah / @cpuguy83 rebased and added one integration test which looks working fine even if builds are still running:
```
21:01:12 === RUN   TestCreateServiceMaxReplicas
21:01:16 --- PASS: TestCreateServiceMaxReplicas (3.84s)
21:01:16     daemon.go:296: [d5086331a5efc] waiting for daemon to start
21:01:16     daemon.go:328: [d5086331a5efc] daemon started
21:01:16     daemon.go:283: [d5086331a5efc] exiting daemon
```

File: integration/service/create_test.go [hunk 3]
```diff
@@ -440,7 +462,7 @@ func TestCreateServiceSysctls(t *testing.T) {
 	}
 }
 
-func serviceRunningTasksCount(client client.ServiceAPIClient, serviceID string, instances uint64) func(log poll.LogT) poll.Result {
+func serviceRunningTasksCount(client client.ServiceAPIClient, serviceID string, instances, maxReplicas uint64) func(log poll.LogT) poll.Result {
```
Comment (cpuguy83) [inline integration/service/create_test.go]: Do we need to change this?
I'm thinking we could do two samples in the test, 1 to make sure that the expected number (max replicas) is reached, and another to make sure that we never get passed max replicas.

Granted checking to make sure that we never go passed the max is racey by definition (e.g. a bad implementation could just have taken longer than expected to place the task)... but I'm not sure how else we'd do that.

Comment (olljanat) [reply integration/service/create_test.go]: @cpuguy83 there is quite many test cases on swarmkit side: https://github.com/docker/swarmkit/blob/master/manager/scheduler/scheduler_test.go#L3266-L3517 is there really point to duplicate them here? IMO, swarmkit is responsible of this logic and on my side we are interested that API works as expected.

Comment (cpuguy83) [reply integration/service/create_test.go]: If we don't want to test that, then can't we only pass MaxReplicas in as `instances` and forgo adding the new argument?

Comment (olljanat) [reply integration/service/create_test.go]: @cpuguy83 changed like that

**Comment by olljanat** [comment]
@cpuguy83 @thaJeztah PTAL, now this contains only simple API test without logic test as it is on swarmkit side.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some minor nits inline (no critical issues, but would be nice to have them in the PR)

I was finally able to give this a spin (see below for some playing with the new option); one thing I noticed (also, see below) that _reducing_ the max-replicas value does not terminate existing tasks. I wondered if this was by design, or an oversight? Not a show-stopper; just curious 🤗 


Here's some things I tried:


Create a service with 2 replicas, but maxreplicas 1

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.40/services/create" \
  -H "Content-Type: application/json" \
  -d '{"EndpointSpec":{"Mode":"vip"},"Labels":{},"Mode":{"Replicated":{"Replicas":2}},"Name":"test","TaskTemplate":{"ContainerSpec":{"DNSConfig":{},"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da"},"ForceUpdate":0,"Placement":{"MaxReplicas":1,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"Resources":{"Limits":{},"Reservations":{}}}}'

{"ID":"eln7vce863tedwd5q72kjaxkg"}
```

Verify the `MaxReplicas` configuration is stored:

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test?insertDefaults=true" | jq .Spec.TaskTemplate.Placement.MaxReplicas

1
```

And check that no more than 1 replica is running (on a single-node swarm);


```bash
docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
qy2qc9p1yder        test                replicated          1/2                 nginx:alpine        
```

Check the task, and confirm it's unable to start due to the `MaxReplicas` restriction:

```bash
docker service ps test --no-trunc --format 'table {{.Name}}\t{{.Error}}'
NAME                ERROR
test.1              
test.2              "no suitable node (max replicas per node limit exceed)"
```

Updated `MaxReplicas` to 2;


Get the service definition

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test"
```
```json
{"ID":"cqnbsmr9foblwucehzu9ekrpb","Version":{"Index":87},"CreatedAt":"2018-12-23T22:23:22.415002133Z","UpdatedAt":"2018-12-23T22:23:22.415002133Z","Spec":{"Name":"test","Labels":{},"TaskTemplate":{"ContainerSpec":{"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da","DNSConfig":{},"Isolation":"default"},"Resources":{"Limits":{},"Reservations":{}},"Placement":{"MaxReplicas":1,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":2}},"EndpointSpec":{"Mode":"vip"}},"Endpoint":{"Spec":{}}}
```

Update the service to set MaxReplicas to 2

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.40/services/test/update?version=87" \
  -H "Content-Type: application/json" \
  -d '{"Name":"test","Labels":{},"TaskTemplate":{"ContainerSpec":{"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da","DNSConfig":{},"Isolation":"default"},"Resources":{"Limits":{},"Reservations":{}},"Placement":{"MaxReplicas":2,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":2}},"EndpointSpec":{"Mode":"vip"}}'
```

Verify the service was updated with the new value (`2`);

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test" | jq .Spec.TaskTemplate.Placement.MaxReplicas

2
```

And confirm that the service was now able to deploy 2 replicas:

```bash
docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
cqnbsmr9fobl        test                replicated          2/2                 nginx:alpine        
```

Now update the service again, and set `MaxReplicas` back to 1


```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.40/services/test/update?version=172" \
  -H "Content-Type: application/json" \
  -d '{"Name":"test","Labels":{},"TaskTemplate":{"ContainerSpec":{"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da","DNSConfig":{},"Isolation":"default"},"Resources":{"Limits":{},"Reservations":{}},"Placement":{"MaxReplicas":1,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":2}},"EndpointSpec":{"Mode":"vip"}}'
```

Verify the service was updated with the new value (`1`)

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test" | jq .Spec.TaskTemplate.Placement.MaxReplicas

1
```

**Note that when _reducing_ the maximum number of replicas, existing tasks are not terminated;**

```bash
docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
cqnbsmr9fobl        test                replicated          2/2                 nginx:alpine        
```


Another thing that needs to be taken into account is that scaling/updating a service with an older version of the CLI will strip the `MaxReplicas`. That's the expected behavior, but perhaps something we should mention somewhere in the documentation :thinking_face:

```bash
docker service scale test=3
test scaled to 3
```

After scaling, the `MaxReplicas` property is reset (because the CLI does not know the new property);

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test?insertDefaults=true" | jq .Spec.TaskTemplate.Placement.MaxReplicas

null
```

**Comment by olljanat** [comment]
@thaJeztah updated based on review. Now hopefully ready to be merged?

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by thaJeztah** [comment]
> By design as we don't want kill running containers when it is not absolutely necessary and user have option to use --force (not sure what it is on API side) and then all tasks will be re-scheduled and situation fixed.

Gotcha; makes sense. We'll have to document that, and explain the behavior (i.e., if they want to force a reschedule, use the `--force`)

Trying that scenario:


Create a service with `MaxReplicas: 2`, and `Replicas: 2`

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.40/services/create" \
  -H "Content-Type: application/json" \
  -d '{"EndpointSpec":{"Mode":"vip"},"Labels":{},"Mode":{"Replicated":{"Replicas":2}},"Name":"test","TaskTemplate":{"ContainerSpec":{"DNSConfig":{},"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da"},"ForceUpdate":0,"Placement":{"MaxReplicas":2,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"Resources":{"Limits":{},"Reservations":{}}}}'

{"ID":"kdu0l3uxhk6p52xa719581h0h"}
```

Verify the `MaxReplicas` configuration is stored:

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test" | jq .Spec.TaskTemplate.Placement

2
```

Confirm the service has two tasks running:

```bash
docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
rmg5iwvxl158        test                replicated          2/2                 nginx:alpine        
```


Update `MaxReplicas` to 2, and set `ForceUpdate` to a new value (to force the update);


Get the service definition

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  "http://localhost/v1.40/services/test"
```
```json
{"ID":"kdu0l3uxhk6p52xa719581h0h","Version":{"Index":27},"CreatedAt":"2018-12-24T09:37:12.633108733Z","UpdatedAt":"2018-12-24T09:37:12.633108733Z","Spec":{"Name":"test","Labels":{},"TaskTemplate":{"ContainerSpec":{"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da","DNSConfig":{},"Isolation":"default"},"Resources":{"Limits":{},"Reservations":{}},"Placement":{"MaxReplicas":2,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":2}},"EndpointSpec":{"Mode":"vip"}},"Endpoint":{"Spec":{}}}
```

Update the service, set `MaxReplicas` to 1, and set `ForceUpdate` to a new value (to force the update);

```bash
curl -s \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.40/services/test/update?version=27" \
  -H "Content-Type: application/json" \
  -d '{"Name":"test","Labels":{},"TaskTemplate":{"ContainerSpec":{"Image":"nginx:alpine@sha256:2e497c294e3ba84aaeab7a0fbb1027819cd1f5f5892ed3c4a82b8b05010090da","DNSConfig":{},"Isolation":"default"},"Resources":{"Limits":{},"Reservations":{}},"Placement":{"MaxReplicas":1,"Platforms":[{"Architecture":"amd64","OS":"linux"},{"Architecture":"arm","OS":"linux"},{"Architecture":"arm64","OS":"linux"},{"Architecture":"386","OS":"linux"},{"Architecture":"ppc64le","OS":"linux"},{"Architecture":"s390x","OS":"linux"}]},"ForceUpdate":1,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":2}},"EndpointSpec":{"Mode":"vip"}}'

{"Warnings":null}
```

Confirm that the service now has 1 task running:


```bash
docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
kdu0l3uxhk6p        test                replicated          1/2                 nginx:alpine      
```

And that this is due to the new `MaxReplicas` limit:

```bash
docker service ps test --no-trunc --format 'table {{.Name}}\t{{.Error}}'
NAME                ERROR
test.1              
test.2              "no suitable node (max replicas per node limit exceed)"
 \_ test.2          
```

Looking good 👍

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!!!

**Comment by olljanat** [comment]
Derek add label: rebuild/powerpc

**Comment by olljanat** [comment]
Derek add label: rebuild/powerpc

**Comment by olljanat** [comment]
Derek add label: rebuild/powerpc

**Comment by thaJeztah** [comment]
@olljanat was the failure only a flaky test? If so, I'm happy to ignore PowerPC and merge

**Comment by olljanat** [comment]
@thaJeztah yes. Those which I disabled on #38423 looks causing issues.

**Comment by vpotseluyko** [comment]
Hi! I greatly love this commit. With what docker & swarm-kit version could it be actually used at production

**Comment by thaJeztah** [comment]
It's not released yet; will be in the next docker release (but you should be able to give it a spin with the nightly builds https://download.docker.com/linux/ubuntu/dists/bionic/pool/nightly/amd64/

**Pull request merged**